packages/pyright-internal/src/common/uri/baseUri.ts

```diff
@@ -0,0 +1,310 @@
+            if (relativeToComponents.length > 0) {
+                return ['.', ...relativeToComponents].join('/');
+            }
+        }
+        return undefined;
+    }
+
+    @cacheMethodWithNoArgs()
+    getPathComponents(): readonly string[] {
+        // Make sure to freeze the result so that it can't be modified.
+        return Object.freeze(this.getPathComponentsImpl());
+    }
+
+    abstract getPath(): string;
+
+    abstract getFilePath(): string;
+
+    @cacheMethodWithArgs()
+    getRelativePathComponents(to: Uri): readonly string[] {
+        const fromComponents = this.getPathComponents();
+        const toComponents = to.getPathComponents();
+
+        let start: number;
+        for (start = 0; start < fromComponents.length && start < toComponents.length; start++) {
+            const fromComponent = fromComponents[start];
+            const toComponent = toComponents[start];
+            if (fromComponent !== toComponent) {
+                break;
+            }
+        }
+
+        if (start === 0) {
+            return toComponents;
+        }
+
+        const components = toComponents.slice(start);
+        const relative: string[] = [];
+        for (; start < fromComponents.length; start++) {
+            relative.push('..');
+        }
+        return [...relative, ...components];
+    }
+
+    getShortenedFileName(maxDirLength: number = 15): string {
+        return getShortenedFileName(this.getPath(), maxDirLength);
+    }
+
+    @cacheMethodWithNoArgs()
+    stripExtension(): Uri {
+        const base = this.filename;
+        const index = base.lastIndexOf('.');
+        if (index > 0) {
+            const stripped = base.slice(0, index);
+            return this.getDirectory().combinePaths(stripped);
+        } else {
+            return this;
+        }
+    }
+
+    @cacheMethodWithNoArgs()
+    stripAllExtensions(): Uri {
+        const base = this.filename;
+        const stripped = base.split('.')[0];
+        if (stripped === base) {
+            return this;
+        } else {
+            return this.getDirectory().combinePaths(stripped);
+        }
+    }
+
+    @cacheMethodWithNoArgs()
+    protected getRoot(): Uri {
+        return this.getRootImpl();
+    }
+    protected abstract getRootImpl(): Uri;
+    protected abstract getRootPath(): string;
+
+    protected abstract getDirectoryImpl(): Uri;
+
+    protected reducePathComponents(components: string[]): string[] {
+        if (!some(components)) {
+            return [];
+        }
+
+        // Reduce the path components by eliminating
+        // any '.' or '..'.
+        const reduced = [components[0]];
+        for (let i = 1; i < components.length; i++) {
+            const component = components[i];
+            if (!component || component === '.') {
+                continue;
+            }
+
+            if (component === '..') {
+                if (reduced.length > 1) {
+                    if (reduced[reduced.length - 1] !== '..') {
```

Comment:
I copied this algorithm from path utils. It does seem like it doesn't eliminate '..' on the front, but I don't think that's possible to have because the front is always the root path. That's why it starts at 1.

I'll add a comment to that effect.